src/iterable/mod.rs

```diff
@@ -0,0 +1,106 @@
+//! A base library for iterfacing with streams of vectors and matrices.
+//!
+//! This library presents the abstraction layer for the _streaming model_.
+//! Essentially, it provides a set of handy utilities as a wrapper around
+//! iterators.
+
+/// The trait [`Iterable`] represents a streamable object that can produce
+/// an arbitrary number of streams of length [`Iterable::len`](Iterable::len).
+///
+/// An Iterable is pretty much like an [`IntoIterator`] that can be copied over
+/// and over, and has an hint of the length.
+/// In other words, these two functions are pretty much equivalent.
+///
+/// # Examples
+///
+/// ```
+/// use ark_std::borrow::Borrow;
+/// use ark_std::iterable::Iterable;
+///
+/// // Relying only on standard library
+/// fn f(xs: impl IntoIterator<Item=impl Borrow<u32>> + Clone) -> u32 {
+///     xs.clone().into_iter().fold(1, |x, y| x.borrow() * y.borrow()) +
+///     xs.clone().into_iter().fold(0, |x, y| x.borrow() + y.borrow()) +
+///     xs.into_iter().size_hint().0 as u32
+/// }
+///
+/// // Relying on the trait below
+/// fn g(xs: impl Iterable<Item=impl Borrow<u32>>) -> u32 {
+///     xs.iter().fold(1, |x, y| x.borrow() * y.borrow()) +
+///     xs.iter().fold(0, |x, y| x.borrow() + y.borrow()) +
+///     xs.len() as u32
+/// }
+///
+/// // Test over a slice (which implements both traits).
+/// let xs = &[1, 2, 3, 4];
+/// assert_eq!(f(xs), g(xs));
+/// ```
+///
+/// # Efficency
+///
+/// For efficiency, functions using iterables are often times relying on
+/// [`Borrow`](std::borrow::Borrow) in order to avoid copying the entire items.
+///
+/// The `Iter` trait itself has a lifetime that is independent from the
+/// [`Iterable`] object. This means that implicitly a copy of the relevant
+/// contents of the object will happen whenever
+/// [`Iterable::iter`](crate::iterable::Iterable::iter) is called. This might
+/// change in the future as associated type constructors [[RFC1598](https://github.com/rust-lang/rfcs/blob/master/text/1598-generic_associated_types.md#declaring--assigning-an-associated-type-constructor)]
+/// stabilize.
+///
+/// # Future implementation
+///
+/// A lot of stream operations must be performed symbolycally.
+/// We expect that, in the future, this trait will accomodate for additional
+/// streaming function, e.g. `Iterable::hadamard(&self, other: &Iterable)` to
+/// perform the hadamard product of two streams, or `Iterable::add(&self, other:
+/// &Iterable)` to perform the addition of two streams.
+pub trait Iterable {
+    /// The type of the element being streamed.
+    type Item;
+    /// The type of the iterator being generated.
+    type Iter: Iterator<Item = Self::Item>;
+
+    ///  Returns the iterator associated to the current instance.
+    ///
+    /// In the so-called _streaming model_, this is equivalent to instantiating
```

Comment:
I know the paper isn't online yet, but is there a way to reference it somehow?

Reply:
Added reference to the paper, we are going to add a link as soon as it gets published.